scib/metrics/morans_i.py

```diff
@@ -0,0 +1,149 @@
+import scanpy as sc
+import numpy as np
+import pandas as pd
+
+from scib.preprocessing import hvg_batch
+
+
+def morans_i(
+    adata_pre,
+    adata_post,
+    batch_key,
+    n_hvg=1000,
+    embed="X_pca",
+    rescale=True,
+    compare_pre=False,
+    hvgs=None,
+):
+    """
+    Compute Moran's I on HVGs (defined across batches) on integrated data
+        as a measure of bio conservation. The metric can be computed either as
+        A.) mean difference between Morans's I on embedding of integrated data and
+        max Moran's I of individual non-integrated batches or
+        B.) mean Morans's I on embedding of integrated data.
+    :param adata_pre: Non-integrated data.
+        Embedding is computed on scaled (m=0, s=1) X (expression)
+        per batch (recomputing HVGs, scaling, pca, neighbours).
+        For Moran's I unscaled X is used.
+    :param adata_post: Integrate data. Should have precomputed embedding and expression in X.
+    :param batch_key: Batch obs col in adata_pre.
+    :param n_hvg: How many top HVGs to use for Moran's I computation.
+    :param embed: Embedding to use for computation of neighbours in adata_post
+    :param rescale: Whether to rescale result to [0,1] so that 1 is better bio conservation
+    :param compare_pre: Whether to compute metric under scenario A instead of B.
+    :param hvgs: Custom list of genes to use for Moran's I computation (instead of hvgs
+        computed across batches).
+    :return: The resulting metric.
+    """
+    # Prepare adatas
+    # Get integrated connectivities for Moran's I
+    adata_post = adata_post.copy()
+    sc.pp.neighbors(adata_post, use_rep=embed)
+    # Prepare pre data
+    adata_pre = adata_pre.copy()
+    adata_pre.obs[batch_key] = adata_pre.obs[batch_key].astype("category")
+
+    # Get HVGs across samples
+    if hvgs is None:
+        hvgs = hvg_batch(
+            adata_pre, batch_key=batch_key, target_genes=n_hvg, flavor="cell_ranger"
+        )
+    else:
+        assert adata_pre.var_names.isin(hvgs).sum() == len(hvgs)
+
+    def compute_mi(data, hvgs):
+        """
+        Helper function for computing Moran's I on HVGS that are non-constant
+        and adding them to var
+        :param data: adata, result is added to var
+        :param hvgs: Genes for which to compute Moran's I
+        """
+        # Make sure that genes used for computation are non-constant
+        cond = (
+            np.array(np.mean(data[:, hvgs].X, axis=0))
+            != np.array(data[0, hvgs].X.todense())
+        ).squeeze()
```

Comment:
mean equal to first value could happen if the genes are not constant, no? Why not define this via standard deviation?

Reply:
You are right! std is better!